pkg/sidecar/sidecar.go

```diff
@@ -183,6 +189,86 @@ func (s *Sidecar) dumpBundles(svidResponse *workloadapi.X509Context) error {
 	return nil
 }
 
+func (s *Sidecar) writeJSON(fileName string, certs map[string]interface{}) {
+	file, err := json.Marshal(certs)
+	if err != nil {
+		s.config.Log.Warnf("Unable to parse certs: %v", err)
+	}
+
+	jsonPath := path.Join(s.config.CertDir, fileName)
+	err = os.WriteFile(jsonPath, file, os.ModePerm)
+	if err != nil {
+		s.config.Log.Warnf("Unable to write JSON file: %v", err)
+	}
+}
+
+func (s *Sidecar) updateJWTBundle(jwkSet *jwtbundle.Set) {
+	s.config.Log.Info("Updating JWK bundles")
+
+	bundles := make(map[string]interface{})
+	for _, bundle := range jwkSet.Bundles() {
+		bytes, err := bundle.Marshal()
+		if err != nil {
+			s.config.Log.Warnf("Unable to marshal JWK bundle: %v", err)
+			continue
+		}
+		bundles[bundle.TrustDomain().Name()] = base64.StdEncoding.EncodeToString(bytes)
+	}
+
+	s.writeJSON(s.config.JWKFilename, bundles)
+}
+
+func (s *Sidecar) fetchJWTSVID(options ...workloadapi.ClientOption) (*jwtsvid.SVID, error) {
+	clientOptions := workloadapi.WithClientOptions(options...)
+
+	jwtSource, err := workloadapi.NewJWTSource(context.Background(), clientOptions)
+	if err != nil {
+		s.config.Log.Warnf("Unable to create JWTSource: %v", err)
+		return nil, err
+	}
+	defer jwtSource.Close()
+
+	jwtSVID, err := jwtSource.FetchJWTSVID(context.Background(), jwtsvid.Params{Audience: s.config.JWTAudience})
+	if err != nil {
+		s.config.Log.Warnf("Unable to fetch JWT SVID: %v", err)
+		return nil, err
+	}
+
+	_, err = jwtsvid.ParseAndValidate(jwtSVID.Marshal(), jwtSource, []string{s.config.JWTAudience})
+	if err != nil {
+		s.config.Log.Warnf("Unable to parse or validate token: %v", err)
+		return nil, err
+	}
+
+	return jwtSVID, nil
+}
+
+func (s *Sidecar) updateJWTSVID(ctx context.Context, options ...workloadapi.ClientOption) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			s.config.Log.Infof("Updating JWT SVID")
+			jwtSVID, err := s.fetchJWTSVID(options...)
+			if err != nil {
+				time.Sleep(time.Second)
```

Comment:
what's the reason for sleeping here?

Reply:
If at half of the expiry time the spiffe helper couldn't get a SVID it should retry it. Without that sleep the helper will make a huge number of requests.

Reply:
I see. Could you use tickers for refreshing and retrying? 

Try something like this:

```
func (s *Sidecar) updateJWTSVID(ctx context.Context, options ...workloadapi.ClientOption) {
    refreshInterval := s.calculateRefreshInterval() 
    retryInterval := 5 * time.Second // This could also be function using a backoff strategy

    // Initialize the ticker with the refresh interval
    ticker := time.NewTicker(refreshInterval)
    defer ticker.Stop()

    for {
        select {
        case <-ctx.Done():
            return
        case <-ticker.C:
            s.config.Log.Infof("Updating or retrying JWT SVID")
            jwtSVID, err := s.fetchJWTSVID(options...)
            if err != nil {
                s.config.Log.Errorf("Failed to fetch JWT SVID: %v", err)
                // Set the ticker to the retry interval
                ticker.Reset(retryInterval)
                continue
            }

            // JWT SVID fetched, now write it to file
            filePath := path.Join(s.config.CertDir, s.config.JWTFilename)
            err = os.WriteFile(filePath, []byte(jwtSVID.Marshal()), os.ModePerm)
            if err != nil {
                s.config.Log.Errorf("Unable to write JWT SVID to a file: %v", err)
                // Keep the ticker at the retry interval
                ticker.Reset(retryInterval)
                continue
            }

            // Update successful, log and set the ticker to the next refresh interval
            s.config.Log.Infof("JWT SVID updated successfully")
            ticker.Reset(refreshInterval)
        }
    }
}
```